### PR TITLE
Add MissionSources and proper RewardDeck system

### PR DESCRIPTION
**Pull request by cannonfodder1**
- [x] Add P1 and P2 Mission Source templates
- [x] Get rid of random mission spawning
- [x] Get rid of failure penalties
- [x] P1s spawn P2s as a reward for success

### CODE REVIEW

**Comment by cannonfodder1** [comment]
TODO: Fix the alert that should show up when P2s are spawned after a successful P1

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc [hunk 1]
```diff
@@ -47,30 +52,78 @@ static protected function bool RewardNotAvaliable(optional XComGameState NewGame
...
-	Template.GiveRewardFn = GiveGatherLeadDarkEvent;
+	Template.GiveRewardFn = GiveGatherLeadPersonnel;
 
 	return Template;
 }
 
 static protected function string GatherLeadString(XComGameState_Reward RewardState)
 {
-	return "Gather lead";
+	return "Gather Lead";
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc]: Should be localized, I used simple string just for testing

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc [hunk 2]
```diff
@@ -47,30 +52,78 @@ static protected function bool RewardNotAvaliable(optional XComGameState NewGame
...
+static protected function GiveGatherLeadTarget(XComGameState NewGameState, XComGameState_Reward RewardState, optional StateObjectReference AuxRef, optional bool bOrder=false, optional int OrderHours=-1)
+{
+	local XComGameState_ResistanceFaction FactionState;
+
+	FactionState = XComGameState_ResistanceFaction(NewGameState.ModifyStateObject(class'XComGameState_ResistanceFaction', FindRandomMetFaction().ObjectID));
+
+	SpawnCovertAction(NewGameState, FactionState, 'CovertAction_P2SupplyRaid');
+	SpawnCovertAction(NewGameState, FactionState, 'CovertAction_P2DarkVIP');
+
+	class'UIUtilities_Infiltration'.static.InfiltrationActionAvaliable(AuxRef, NewGameState);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc]: The first argument is the action to focus when opening the screen. In this case `AuxRef.ObjectID` is always `0` (you can check the mission source code)

I suggest that you make the first argument in `InfiltrationActionAvaliable` optional and just omit it here

File: CovertInfiltration/Src/CovertInfiltration/Classes/CI_XComGameState_CovertAction.uc [hunk 3]
```diff
@@ -153,4 +153,32 @@ function RemoveEntity(XComGameState NewGameState)
...
+
+function string GetDarkEventString()
+{
+	local XComGameState_Reward RewardState;
+	local XComGameState_DarkEvent DarkEventState;
+
+	RewardState = XComGameState_Reward(`XCOMHISTORY.GetGameStateForObjectID(RewardRefs[0].ObjectID));
+	DarkEventState = XComGameState_DarkEvent(`XCOMHISTORY.GetGameStateForObjectID(RewardState.RewardObjectReference.ObjectID));
+
+	return DarkEventState.GetMyTemplate().DisplayName;
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/CI_XComGameState_CovertAction.uc]: Check for none - not all CAs have an associated DE

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc [hunk 4]
```diff
@@ -95,11 +95,32 @@ static protected function string GatherLeadString(XComGameState_Reward RewardSta
...
+	NewRewardState = XComGameState_Reward(`XCOMHISTORY.GetGameStateForObjectID(ActionState.RewardRefs[0].ObjectID));
+	NewRewardState.RewardObjectReference = DarkEvents[0];
+	
+	if(DarkEvents.Length == 1)
+	{
+		`RedScreen("CI: ONLY ONE DARK EVENTS TO COUNTER, CANNOT SPAWN SECOND P2");
+		return;
+	}
+
+	ActionState = SpawnCovertAction(NewGameState, FactionState, 'CovertAction_P2DarkEvent');
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc]: Please use a loop -.-

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc [hunk 5]
```diff
@@ -95,11 +95,32 @@ static protected function string GatherLeadString(XComGameState_Reward RewardSta
...
-	SpawnCovertAction(NewGameState, FactionState, 'CovertAction_P2DarkEvent');
-	SpawnCovertAction(NewGameState, FactionState, 'CovertAction_P2DarkEvent');
+	if(DarkEvents.Length == 0)
+	{
+		`RedScreen("CI: NO PENDING DARK EVENTS TO COUNTER, CANNOT SPAWN P2s");
+		return;
+	}
+
+	ActionState = SpawnCovertAction(NewGameState, FactionState, 'CovertAction_P2DarkEvent');
+	NewRewardState = XComGameState_Reward(`XCOMHISTORY.GetGameStateForObjectID(ActionState.RewardRefs[0].ObjectID));
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc]: Fetch from `NewGameState`, not history

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc [hunk 6]
```diff
@@ -151,19 +172,56 @@ static function StateObjectReference FindRandomMetFaction()
...
+	local StateObjectReference ActionRef;
 	
 	StrategyTemplateManager = class'X2StrategyElementTemplateManager'.static.GetStrategyElementTemplateManager();
 	ActionTemplate = X2CovertActionTemplate(StrategyTemplateManager.FindStrategyElementTemplate(ActionTemplateName));
 	
-	FactionState.AddCovertAction(NewGameState, ActionTemplate, ExclusionList);
+	ActionRef = FactionState.CreateCovertAction(NewGameState, ActionTemplate, eFactionInfluence_Minimal);
+	FactionState.CovertActions.AddItem(ActionRef);
+
+	ActionState = XComGameState_CovertAction(`XCOMHISTORY.GetGameStateForObjectID(ActionRef.ObjectID));
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc]: Fetch from `NewGameState`, not history

File: CovertInfiltration/Config/XComInfiltration.ini [hunk 7]
```diff
@@ -183,10 +183,9 @@ WorkRequiredForP1Variance[1]=33000
 WorkRequiredForP1Variance[2]=33000
 WorkRequiredForP1Variance[3]=33000
 
-; Defines the pool of actions (templates) that will be used to choose from
-; Current values are for testing only while #63 is implemented
-ActionsToSpawn=CovertAction_PrepareRaid
-;ActionsToSpawn=CovertAction_RecruitEngineer
+; If the user doesn't have enough of these resources, a P1SupplyRaid will be spawned on the next P1 spawn
```
Comment (Xymanek) [inline CovertInfiltration/Config/XComInfiltration.ini]:
```
... provided it wasn't spawned just before
```

File: CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_MissionSiteInfiltration.uc [hunk 8]
```diff
@@ -39,10 +39,14 @@ function SetupFromAction(XComGameState NewGameState, XComGameState_CovertAction
 		MissionRewards.AddItem(RewardState);
 	}
 
+	ActionReward = XComGameState_Reward(`XCOMHISTORY.GetGameStateForObjectID(Action.RewardRefs[0].ObjectID));
+	DarkEvent = ActionReward.RewardObjectReference;
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_MissionSiteInfiltration.uc]: Check if reward template is the DE one

**Pull request merged**